File: 2023/day2/main.py
import re


def main():
    gid_sum = 0
    power_sum = 0
    with open('input.txt') as f:
        for line in f:
            gid, sets = parse_game(line)
            min_game = get_minimum_set(sets)
            power_sum += min_game['red'] * min_game['green'] * min_game['blue']
            if is_game_valid(sets):
                gid_sum += gid
    print(f"{gid_sum=}")
    print(f"{power_sum=}")


def parse_game(line):
    sets = []
    pattern = r"Game (\d+): (.*)"
    m = re.match(pattern, line)
    gid, rest = m.groups()
    for gameset in map(str.strip, rest.split(';')):
        d = {}
        for cubes in map(str.strip, gameset.split(',')):
            num_cubes, color = cubes.split()
            d[color] = int(num_cubes)
        sets.append(d)
    return int(gid), sets


def is_game_valid(sets):
    return all((is_set_valid(gameset) for gameset in sets))


def is_set_valid(gameset):
    max_values = {'red': 12, 'green': 13, 'blue': 14}
    return all((n <= max_values[color] for color, n in gameset.items()))


def get_minimum_set(sets):
    max_values = {'red': 0, 'green': 0, 'blue': 0}
    for gameset in sets:
        for color, n in gameset.items():
            if n > max_values[color]:
                max_values[color] = n
    return max_values


if __name__ == '__main__':
    main()
    # print(parse_game("Game 10: 2 green, 3 red; 18 blue, 20 green, 9 red; 7 red, 9 blue, 17 green")) # noqa

File: 2023/day2/main.go
package main

import (
    "os"
    "log"
    "bufio"
    "regexp"
    "strings"
    "strconv"
)


func main() {
    file, err := os.Open("input.txt")
    if err != nil {
        log.Fatal(err)
    }
    defer file.Close()
    
    sum := 0
    sumPower := 0
    scanner := bufio.NewScanner(file)

    for scanner.Scan() {
        line := scanner.Text()
        gid, game := parseGame(line)
        if isGamePossible(game) {
            sum += gid
        }
        minSet := getMinSet(game)
        sumPower += minSet["red"] * minSet["green"] * minSet["blue"]
    }

    if err := scanner.Err(); err != nil {
        log.Fatal("Invalid input: %s", err)
    }

    log.Printf("result1=%d", sum)
    log.Printf("result2=%d", sumPower)
}

func parseGame(line string) (int, []map[string]int) {
    game := []map[string]int{}
    
    pattern := regexp.MustCompile(`Game (\d+): (.*)`)
    match := pattern.FindStringSubmatch(line)

    gid, _ := strconv.Atoi(match[1])
    rest := match[2]

    for _, gameset := range strings.Split(rest, ";") {
        gameset = strings.TrimSpace(gameset)
        d := map[string]int{}

        for _, cube := range strings.Split(gameset, ",") {
            cube = strings.TrimSpace(cube)
            num_and_color := strings.Split(cube, " ")
            num, _ := strconv.Atoi(num_and_color[0])
            color := num_and_color[1]
            d[color] = num
        }
        game = append(game, d)
    }

    return gid, game
}

func isGamePossible(game []map[string]int) bool {
    maxCubes := map[string]int{
        "red": 12,
        "green": 13,
        "blue": 14,
    }
    for _, cubes := range game {
        for color, num := range cubes {
            if maxCubes[color] < num {
                return false
            }
        }
    }
    return true
}

func getMinSet(game []map[string]int) map[string]int {
    maxCubes := map[string]int{
        "red": 0,
        "green": 0,
        "blue": 0,
    }
    for _, gameset := range game {
        for color, count := range gameset {
            if count > maxCubes[color] {
                maxCubes[color] = count
            }
        }
    }
    return maxCubes
}
